File: go/internal/agent/gemini_cli.go
package agent

import (
	"context"
	"fmt"
	"time"
)

// GeminiCliAgent represents the Go port of the Gemini CLI functionality
type GeminiCliAgent struct {
	BaseAgent
	UseSearchGrounding bool
	Checkpoints        map[string]string
}

// NewGeminiCliAgent initializes a new Gemini CLI agent
func NewGeminiCliAgent() *GeminiCliAgent {
	return &GeminiCliAgent{
		UseSearchGrounding: true,
		Checkpoints:        make(map[string]string),
	}
}

// GenerateWithGrounding queries the agent with Google Search grounding enabled
func (a *GeminiCliAgent) GenerateWithGrounding(ctx context.Context, prompt string) (string, error) {
	time.Sleep(400 * time.Millisecond)
	groundingInfo := "Searched Google for: latest context"
	if !a.UseSearchGrounding {
		groundingInfo = "Grounding disabled"
	}
	return fmt.Sprintf("Response for '%s' [%s]", prompt, groundingInfo), nil
}

// SaveCheckpoint saves the current conversation state
func (a *GeminiCliAgent) SaveCheckpoint(checkpointName string, state string) error {
	a.Checkpoints[checkpointName] = state
	return nil
}

// LoadCheckpoint restores the conversation state
func (a *GeminiCliAgent) LoadCheckpoint(checkpointName string) (string, error) {
	if state, ok := a.Checkpoints[checkpointName]; ok {
		return state, nil
	}
	return "", fmt.Errorf("checkpoint not found: %s", checkpointName)
}

// StreamJSON simulates JSON streaming output
func (a *GeminiCliAgent) StreamJSON(ctx context.Context, prompt string) (<-chan string, error) {
	ch := make(chan string)
	go func() {
		defer close(ch)
		for i := 0; i < 3; i++ {
			select {
			case <-ctx.Done():
				return
			default:
				ch <- fmt.Sprintf(`{"chunk": %d, "content": "part %d"}`, i, i)
				time.Sleep(100 * time.Millisecond)
			}
		}
	}()
	return ch, nil
}

File: rust/src/gemini_cli.rs
use std::collections::HashMap;
use std::time::Duration;
use tokio::sync::mpsc;
use tokio::time::sleep;

pub struct GeminiCliAgent {
    pub use_search_grounding: bool,
    pub checkpoints: HashMap<String, String>,
}

impl GeminiCliAgent {
    pub fn new() -> Self {
        Self {
            use_search_grounding: true,
            checkpoints: HashMap::new(),
        }
    }

    pub async fn generate_with_grounding(&self, prompt: &str) -> String {
        sleep(Duration::from_millis(400)).await;
        let grounding = if self.use_search_grounding {
            "Searched Google for: latest context"
        } else {
            "Grounding disabled"
        };
        format!("Response for '{}' [{}]", prompt, grounding)
    }

    pub fn save_checkpoint(&mut self, name: &str, state: &str) {
        self.checkpoints.insert(name.to_string(), state.to_string());
    }

    pub fn load_checkpoint(&self, name: &str) -> Option<String> {
        self.checkpoints.get(name).cloned()
    }

    pub async fn stream_json(&self, prompt: &str) -> mpsc::Receiver<String> {
        let (tx, rx) = mpsc::channel(10);
        tokio::spawn(async move {
            for i in 0..3 {
                let msg = format!(r#"{{"chunk": {}, "content": "part {}"}}"#, i, i);
                if tx.send(msg).await.is_err() {
                    break;
                }
                sleep(Duration::from_millis(100)).await;
            }
        });
        rx
    }
}
